Decouple backing off on repeaters from iterating over repeaters

gherceg opened the pull request:
## Product Description
<!-- Where applicable, describe user-facing effects and include screenshots. -->

## Technical Summary
<!--
    Provide a link to any tickets, design documents, and/or technical specifications
    associated with this change. Describe the rationale and design decisions.
-->
https://dimagi.atlassian.net/browse/SAAS-18322

Currently, there is one feature flag, `PROCESS_REPEATERS`, that allows us to seamlessly switch from our current mechanism for queueing and processing repeat records to the new mechanism. My understanding is that in order to ensure a smooth transition between turning this feature flag on and off, we also continue to update repeat records as if `PROCESS_REPEATERS` weren't enabled.

My goal is to get the improved queueing that comes with `PROCESS_REPEATERS` enabled on production without the unknowns that exist with backing off on repeaters instead of repeat records. We are still working through backing off repeaters, since it is a much harder decision to say "we are not going to process any records for this repeater for the next X hours" as opposed to "we are not going to process this specific repeat record for the next X hours".

My understanding is that the main thing that needs to change in order to support processing repeaters but still backing off on repeat records is the order in which we pull repeat records to process for a repeater. Previously, we would always process the oldest repeat records for that repeater, until they either succeed or are cancelled. It seems easy enough to instead sort by `next_attempt` since that is always set when a RepeatRecord is [first created](https://github.com/dimagi/commcare-hq/blob/a8277af9d8a15b05e14e7a041e8a50941cd5afb1/corehq/motech/repeaters/models.py#L492-L492), and updated when the record is processed.

I don't want this to take away from the end goal of backing off at the repeater level since that is ultimately more efficient, but we also need a change to how we process repeat records and it seems that backing off at the repeater level is the only thing holding us back.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
`BACKOFF_REPEATERS` and `PROCESS_REPEATERS`
## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
Maybe? This would really be intended to get onto production asap, so we may be better off testing ourselves.

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

gherceg committed:
Create separate toggle for backing off repeaters

gherceg committed:
Check for PROCESS_REPEATERS and BACKOFF_REPEATERS

Since BACKOFF_REPEATERS isn't actually useful if PROCESS_REPEATERS isn't
also enabled, it makes sense to check for both.

gherceg committed:
Put code behind BACKOFF_REPEATER feature flag

kaapstorm commented on corehq/motech/repeaters/models.py:
```diff
@@ -409,13 +409,18 @@ def _repeater_type(cls):
+        If BACKOFF_REPEATERS is enabled, this will return in the order they
+        were registered, otherwise it will order by the next check date
         """
+        if toggles.BACKOFF_REPEATERS.enabled(self.domain, namespace=toggles.NAMESPACE_DOMAIN):
```
Nicely spotted @millerdev .

Yeah, it's fine. The order of the repeat records does not matter to `CaseRepeater.register()`. The point is that if a given case already has a repeat record (or more) it won't register a new one. `register()` can return any repeat record for the case.

kaapstorm replied:
PR https://github.com/dimagi/commcare-hq/pull/36634

millerdev replied:
Is it now safe to add an assertion in this if-block?
```suggestion
        if toggles.BACKOFF_REPEATERS.enabled(self.domain, namespace=toggles.NAMESPACE_DOMAIN):
            assert toggles.PROCESS_REPEATERS.enabled(self.domain, namespace=toggles.NAMESPACE_DOMAIN)
```

millerdev replied:
I wanted to confirm that the documented behavior is correct. If it is the assertion should be safe.

> The docs for `BACKOFF_REPEATERS` say "This feature flag is only relevant if PROCESS_REPEATERS is also enabled."

I didn't find it trivial to verify that by looking at the code. Should we change the `BACKOFF_REPEATERS` docs if we are not comfortable adding the assertion?

gherceg committed:
Merge branch 'master' into gh/repeaters/backoff-repeater-ff

kaapstorm committed:
repeat_records_ready: next_check <= now

gherceg committed:
Merge branch 'master' into gh/repeaters/backoff-repeater-ff

kaapstorm committed:
Fix CaseRepeater.register() filter

kaapstorm committed:
Fix tests

kaapstorm committed:
Merge branch 'master' into gh/repeaters/backoff-repeater-ff

gherceg committed:
Check both process and backoff flags

Since BACKOFF_REPEATERS is only effective if PROCESS_REPEATERS is
enabled, enforce that in code by checking that both are enabled.

gherceg committed:
Add tests asserting check_repeaters vs process_repeaters

gherceg commented:
Tested with a small case load on staging and saw the expected behavior with the PROCESS_REPEATERS flag enabled.
